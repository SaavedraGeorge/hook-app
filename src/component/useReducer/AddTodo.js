import React from 'react';
import { useForm } from '../../hooks/useForm';

export const AddTodo = ({handleAddTodo}) => {
    const [{description}, handleInputChange, reset] = useForm({ 
        description: ''
     });
    
    const handleSubmit = (e) => {
        e.preventDefault();

        if( description.trim().length <= 1){
            return
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false 
        }

        handleAddTodo(newTodo);
        reset();
    }

    return (
        <>
            <h4>Agregar To Do</h4>
            <hr />
            <form onSubmit={ handleSubmit }>
                <input 
                    type="text" 
                    name="description" 
                    placeholder="add To Do" 
                    autoComplete='off'
                    className='form-control'
                    value={ description }
                    onChange={ handleInputChange }
                />
                <button 
                    className='btn btn-outline-primary col-12 mt-1'
                    type='submit'
                >Agregar</button>
            </form>
        </>
    );
};
