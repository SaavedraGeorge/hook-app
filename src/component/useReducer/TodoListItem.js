import React from 'react';

export const TodoListItem = ({todo, i, handleDelete, toggle }) => {
  
  return (
    <li
    className='list-group-item'
    key={todo.id}
    >
        <p 
            onClick={ () => toggle(todo.id)}
            className={`${ todo.done ? 'complete': ''}`}
        >
            {i+1}. { todo.desc }
        </p>
        <button 
            className='btn btn-danger'
            onClick={ () => handleDelete(todo.id) }
        >Eliminar</button>
    </li>
  );
};
