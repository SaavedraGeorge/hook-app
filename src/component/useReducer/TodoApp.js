import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';


import './style.css'
import { TodoList } from './TodoList';
import { AddTodo } from './AddTodo';

const init = () =>{
    
    // llama lo que hay en el loca storrage y si es vacio devuelve []
    return JSON.parse(localStorage.getItem('todos')) || [];
    
} 

export const TodoApp = () => {

    const [todos, dispatch] = useReducer( todoReducer, [], init );

    
    
    
     useEffect(() => {
       localStorage.setItem('todos', JSON.stringify(todos));
     
     }, [todos]);
     

    const handleAddTodo = ( newTodo ) => {
        
        dispatch({
            type: 'add',
            payload: newTodo
        });

    }

    const handleDelete = (idItem) => {
        
        const action ={
            type: 'delete',
            payload: idItem
        }

        dispatch(action);

    }

    const toggle = (idItem) => {
        dispatch( {
            type: 'toggle',
            payload: idItem 
        } )
    }

    return (
        <div>
            <h1>To Do App <small>{todos.length}</small></h1>
            <hr />

  

            <div className='row'>
                <div className='col-7'>
                    <TodoList 
                        todos = {todos}
                        handleDelete = {handleDelete}
                        toggle = {toggle}
                    /> 
                </div>
                <div className='col-5'>
                    <AddTodo handleAddTodo={handleAddTodo} />
                </div>
            </div>
        </div>
    )
}
