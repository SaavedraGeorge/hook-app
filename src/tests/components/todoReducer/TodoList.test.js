import { shallow } from 'enzyme';
import React from 'react';
import { TodoList } from '../../../component/useReducer/TodoList';
import {demoTodo} from '../../fixtures/demoTodo';

;


describe('pruebas a <TodoList />', () => {
    const handleDelete = jest.fn();
    const toggle = jest.fn();

    const wrapper = shallow( 
    <TodoList 
        todos = { demoTodo }
        handleDelete = { handleDelete }
        toggle = { toggle } 
    />); 

    test('should show correctly component', () => {
        expect( wrapper ).toMatchSnapshot();
    });
    
    test('ver cuantos elementos se renderizan', () => {
        
        expect(wrapper.find('TodoListItem').length).toBe(2);
        //busca en la propiedad del elemento 0 de los todoListItem, si existe handdelete y si es una funcion
        expect(wrapper.find('TodoListItem').at(0).prop('handleDelete')).toEqual(expect.any(Function));
    });
});
