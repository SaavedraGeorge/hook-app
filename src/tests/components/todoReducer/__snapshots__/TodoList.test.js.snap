// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas a <TodoList /> should show correctly component 1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoListItem
    handleDelete={[MockFunction]}
    i={0}
    key="1"
    todo={
      Object {
        "desc": "Aprender React",
        "done": false,
        "id": 1,
      }
    }
    toggle={[MockFunction]}
  />
  <TodoListItem
    handleDelete={[MockFunction]}
    i={1}
    key="2"
    todo={
      Object {
        "desc": "Aprender mongo",
        "done": false,
        "id": 2,
      }
    }
    toggle={[MockFunction]}
  />
</ul>
`;
